File: backend/app/models/appointment.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import ForeignKey, String, Text, Enum, DateTime, Boolean, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import UUIDBase


class AppointmentStatus(str, enum.Enum):
    REQUESTED = "requested"
    CONFIRMED = "confirmed"
    REMINDER_SENT = "reminder_sent"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    NO_SHOW = "no_show"


class Appointment(UUIDBase):
    __tablename__ = "appointments"

    business_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("businesses.id"), index=True
    )
    customer_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("customers.id"), index=True
    )
    conversation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("conversations.id"), nullable=True
    )
    service_name: Mapped[str] = mapped_column(String(255))
    appointment_ref: Mapped[str] = mapped_column(String(20), unique=True, index=True)
    status: Mapped[AppointmentStatus] = mapped_column(
        Enum(AppointmentStatus), default=AppointmentStatus.REQUESTED
    )
    scheduled_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    duration_mins: Mapped[int] = mapped_column(Integer, default=60)
    notes: Mapped[str | None] = mapped_column(Text)

    # Reminder tracking
    reminder_24h_sent: Mapped[bool] = mapped_column(Boolean, default=False)
    reminder_2h_sent: Mapped[bool] = mapped_column(Boolean, default=False)
    followup_sent: Mapped[bool] = mapped_column(Boolean, default=False)

    business: Mapped["Business"] = relationship(back_populates="appointments")
    customer: Mapped["Customer"] = relationship()


class ServiceItem(UUIDBase):
    """Services or products a business offers."""
    __tablename__ = "service_items"

    business_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("businesses.id"), index=True
    )
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text)
    price: Mapped[int] = mapped_column(Integer, default=0)
    currency: Mapped[str] = mapped_column(String(3), default="NGN")
    duration_mins: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    category: Mapped[str | None] = mapped_column(String(100))

    business: Mapped["Business"] = relationship(back_populates="services")

File: mobile/lib/models/appointment.dart
class Appointment {
  final String id;
  final String? customerName;
  final String? customerPhone;
  final String? serviceName;
  final String status;
  final String? scheduledAt;
  final String? notes;
  final String createdAt;

  Appointment({
    required this.id,
    this.customerName,
    this.customerPhone,
    this.serviceName,
    this.status = 'pending',
    this.scheduledAt,
    this.notes,
    required this.createdAt,
  });

  factory Appointment.fromJson(Map<String, dynamic> json) => Appointment(
        id: json['id'] as String,
        customerName: json['customer_name'] as String?,
        customerPhone: json['customer_phone'] as String?,
        serviceName: json['service_name'] as String?,
        status: json['status'] as String? ?? 'pending',
        scheduledAt: json['scheduled_at'] as String?,
        notes: json['notes'] as String?,
        createdAt: json['created_at'] as String,
      );
}

class ServiceItem {
  final String id;
  final String name;
  final String? description;
  final double price;
  final int? durationMinutes;
  final String? category;
  final bool isActive;

  ServiceItem({
    required this.id,
    required this.name,
    this.description,
    this.price = 0,
    this.durationMinutes,
    this.category,
    this.isActive = true,
  });

  factory ServiceItem.fromJson(Map<String, dynamic> json) => ServiceItem(
        id: json['id'] as String,
        name: json['name'] as String,
        description: json['description'] as String?,
        price: (json['price'] as num?)?.toDouble() ?? 0,
        durationMinutes: json['duration_minutes'] as int?,
        category: json['category'] as String?,
        isActive: json['is_active'] as bool? ?? true,
      );
}
